### ISSUE TITLE
[BH] ttnn.repeat has zeros in output

### ISSUE THREAD

**Issue by s-jovic** [tenstorrent/tt-metal#19722]
In SD model we had bad PCC because of `ttnn.repeat` op which leaves some zeros in the output. This is the unittest showing the failing case, however, it usually passes when called by itself:

```
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

def test_mul(device):
    scale = torch.full((1, 1, 1, 1), 0.0791)
    scale_memory = ttnn.DRAM_MEMORY_CONFIG

    ttnn_scale = ttnn.from_torch(scale, device=device, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT, memory_config=scale_memory)
    repeated = ttnn.repeat(ttnn_scale, [1, 16, 1, 1])

    torch_repeated = scale.repeat(1, 16, 1, 1)

    _, message = assert_with_pcc(ttnn.to_torch(repeated), torch_repeated, pcc=0.99)
    print(message)
```

It is always reproducible in a SD unit test on branch `sjovic/sd-mul-repeat` (on main we have a workaround, so can't reproduce):

```
"models/demos/wormhole/stable_diffusion/tests/test_cross_attention.py::test_cross_attention_512x512[input_shape4-shard_layout4-shard_end_core4-shard_shape4-160-down-2-0-False-CompVis/stable-diffusion-v1-4-device_params0]"
```

The printed tensor before the breakpoint is hit has zeros inside - this is the output of ttnn.multiply op that calls ttnn.repeat under the hood for the right operand and that is the source of the problem.

**Comment by s-jovic** [comment]
> the test passes. Is that expected?

Yes, because the PCC margin is 0.99. The PCC of this whole component is 0.997, but when multiply is switched with bcast op (current workaround), PCC is pushed to 0.9986 - this pushes model PCC from 0.994 to 0.996 and this results in big improvement in output images.

This is why I added a breakpoint and printout of the tensor after multiply op - because there the problem is clearly visible since half of the tensor is 0. With the fix this printout should give a tensor with various non-zero values, _and_ the test should still be passing. It's very cumbersome but the problem cannot be reliably reproduced in a unittest unfortunately.


> Are you certain that the repeat op is outputting erroneous 0s 

I am pretty sure, to confirm, I also tried calling `ttnn.repeat(tensor, [1,16,1,1])` before multiply and the output was a vector of size 16 with half zeros.

**Comment by amorrisonTT** [comment]
This is very helpful, thank you

**Comment by s-jovic** [comment]
Yes, this is BH specific issue, the directory is a bit misleading since we didn't do the refactoring yet.

**Comment by amorrisonTT** [comment]
Thanks for clarifying. I'm working on it.

**Comment by amorrisonTT** [comment]
Summary of findings so far:

- Cannot repro in an isolated unit test, but issue is deterministic within that model and can be caught with asserts etc
- Draft PR solves the problem in the issue but root cause is not understood
- Erroneous 0s originated from a `ttnn::typecast` within the `repeat` op, converting from back from bf16->bf_8b after repeat execution

Communicated to @patrickroberts

**Comment by KalaivaniMCW** [comment]
Hi @amorrisonTT , 
I tested on the repro branch amorrison/19722-typecast-zeros-bh-repro

In this case, The issue was that the tensor before being typecasted ( bf16) is fine but after being typecasted to bfloat8_b it has zeros instead of the expected input value.. 

**The tensor before typecast - Logical Shape([1, 16, 1, 1]) - Shape([1, 16, 32, 32])** 
`ttnn.Tensor([[[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]]]], shape=Shape([1,16,1,1]), dtype=DataType::BFLOAT16, layout=Layout::TILE)`

**after typecast - Shape([1, 16, 1, 1]) - Shape([1, 16, 32, 32])** 
`ttnn.Tensor([[[[0.07910]],[[0.00000]],[[0.00000]],[[0.00000]],[[0.00000]],[[0.07910]],[[0.00000]],[[0.00000]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.07910]],[[0.00000]]]], shape=Shape([1,16,1,1]), dtype=DataType::BFLOAT8_B, layout=Layout::TILE)`

The problem seems to occur somewhere in the Repeat operation ( either that or from the RM to Tile layout conversion )

Before Repeat operation we typecast bf8_b input to bf16 
- During this typecast from bfloat8_b to bfloat16 :

I printed the input & output CB when the tensor was of Shape([1, 1, 1, 1]). Looks fine !
```
0:(x=0,y=0):TR0: 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0:(x=0,y=0):TR0: 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```
- After this with the bf16 tensor, it is converted to RM -  repeat operation happens and then it is converted back to TILE layout. 
The tensor goes from of Shape([1, 1, 1, 1]) to Shape([1, 16, 1, 1])

- Next the bfloat16 tensor is converted to bfloat8_b: 
During typecast from bfloat16 to bfloat8_b, I printed the input & output CB when the tensor was of Shape([1, 16, 1, 1]) 
We have a total of 16 tiles for this shape which gives the 16 values we see in the output. I am adding the first few here. 

So some of the input tiles have `inf` and really large numbers in the padding. 
For `bfloat8_b` dtype which has a shared exponent, such really large values will cause nearby values to be flushed to 0 
```
Tile id 0:   (input retained)				
in_CB 0:(x=0,y=0):TR0: 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0	
out_CB 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

Tile id 1: (input flushed)
in_CB 0:(x=0,y=1):TR0: 0.0791016 0 0 0 0 0 0 0 inf 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 3.38953e+38
out_CB: 0 0 0 0 0 0 0 0 inf 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 3.37624e+38

Tile 2: (input flushed)
in_CB 0:(x=0,y=2):TR0: 0.0791016 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 3.38953e+38
out_CB 0:(x=0,y=2):TR0: 0 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 3.37624e+38

Tile id 3: (input flushed)
in_CB 0:(x=0,y=3):TR0: 0.0791016 0 0 0 0 0 0 0 inf 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 3.38953e+38
out_CB 0:(x=0,y=3):TR0: 0 0 0 0 0 0 0 0 inf 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 3.37624e+38

Tile id 4: (input flushed)
in_CB 0:(x=0,y=4):TR0: 0.0791016 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 inf 0 0 0 0 0 0 8.47383e+37
out_CB 0:(x=0,y=4):TR0: 0 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 inf 0 0 0 0 0 0 8.50706e+37

Tile id 5: (input retained)
in_CB 0:(x=0,y=5):TR0: 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 0 3.38953e+38 0 0 0 0 0 0 3.38953e+38
out_CB  0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 0 3.37624e+38 0 0 0 0 0 0 3.37624e+38
```
This is the behaviour I observed on BH when running 
`pytest "models/demos/wormhole/stable_diffusion/tests/test_cross_attention.py::test_cross_attention_512x512[input_shape4-shard_layout4-shard_end_core4-shard_shape4-160-down-2-0-False-CompVis/stable-diffusion-v1-4-device_params0]"`

This doesn't happen in Wormhole.

The workaround you mentioned `working_tensor = ttnn::fill_implicit_tile_padding(working_tensor, 0);` will get rid of the inf/,large values in the padding as it resets to 0, making it work

```
Ex: Tile id 4 (after resetting padding) 
in_CB 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
out_CB 0.0791016 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```

**Comment by amorrisonTT** [comment]
Thank you for the investigation. The nature of the root of the issue, flushing due to garbage values, suggests that the erroneous zeros could show up non-deterministically at any stage in the Op flow. I had some pretty ironclad asserts in the code that showed them showing up in the `typecast` but it seems reasonable that they could show up at other points due to the same issue. Either way, this is a great confirmation of the root cause and justification for my fix, or something similar, which I can take on from here. Thanks again.

**Comment by KalaivaniMCW** [comment]
Sure @amorrisonTT. Although I've seen it happen in other ops like reciprocal for specific inputs, didn't expect padding values to affect the input. It was a nice opportunity to confirm it at the buffer level.

**Comment by scardozaTT** [comment]
**Steps to reproduce the issue**
Modify `models/demos/wormhole/stable_diffusion/tt/ttnn_functional_cross_attention.py`
Replace:
```python
attention_scores = ttnn.bcast(
    attention_scores,
    self.scale,
    math_op=ttnn.BcastOpMath.MUL,
    dim=ttnn.BcastOpDim.HW,
)
```
With:
```python
attention_scores_temp = attention_scores

repeat_scale = ttnn.repeat(self.scale, [1, 16, 1, 1])
assert not (ttnn.to_torch(repeat_scale) == 0).any()
attention_scores = ttnn.multiply(
    attention_scores_temp,
    repeat_scale,
    memory_config=attention_scores.memory_config(),
)

# print(ttnn.to_torch(attention_scores))
# breakpoint()
attention_scores_temp.deallocate()
```
Run `pytest models/demos/wormhole/stable_diffusion/tests/test_cross_attention.py::test_cross_attention_512x512[input_shape4-shard_layout4-shard_end_core4-shard_shape4-160-down-2-0-False-device_params0]`.
(if this pytest config isn't available simply run all of them in that file, some of them will trip the assert)

I'll submit a fix soon: https://github.com/tenstorrent/tt-metal/pull/31609
